### ISSUE TITLE
Gateway throws `This data graph is missing a valid configuration. Cannot read property 'map' of undefined` error when services has not registered any @keys

### ISSUE THREAD

**Issue by NickTomlin** [apollographql/federation#368]
I recently sunk several hours debugging an issue registering an individual service into my unmanaged `@apollo/gateway` server. My gateway logged the following error

```text
ERROR - This data graph is missing a valid configuration. Cannot read property 'map' of undefined
```

A search for that text leads to [this line](https://github.com/apollographql/apollo-server/blob/570f548b88750a06fbf5f67a4abe78fb0f870ccd/packages/apollo-server-core/src/ApolloServer.ts#L438) which purposefully hides the exception.

dropping a breakpoint and inspecting the error, the stack points to [this line](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-federation/src/service/printComposedSdl.ts#L226).

After setting a breakpoint here, I can see that `keys` was null, because my federated service did not have any `@key`s defined.
Understanding that, this error makes sense, but the path here was a bit difficult for someone new to federation.

If this needs to be a failure, would it be possible to explicitly check for this case and provide a warning message that includes the name of the service?

```text
This data graph is missing a valid configuration: `my-great-service` has not provided any `@keys` see [doc link] for more info.`
```

Alternately, the work in apollographql/apollo-server#4136 might allow these to pass through just fine.

Dependencies
---

```
@apollo/gateway@0.19.0
@apollo/server@3.0.0-alpha.1
@apollo/federation@0.19.0
```

Reproduction
---

[repository that reproduces issue](https://github.com/NickTomlin/apollo-gateway-key-error)

**Comment by HunderlineK** [comment]
Generally, better error messages around this issue could be helpful. We have a gateway with multiple providers and each time a user receives an error like this, debugging it will be only possible by people who have access to server logs. Thanks!

**Comment by glasser** [comment]
I don't know if we've improved the root cause here, but error handling on gateway startup is better as of Apollo Server v2.22. As long as you're on that version, error stacktraces should be shown more clearly. Specifically, if you're using `apollo-server` and calling `.listen`, your `listen` call will reject with the startup error if loading the schema fails. If you're using a different package like `apollo-server-express` you can `await server.start()` right after constructing the server and that will throw the startup error if loading the schema fails. Finally, even if you don't do that and start your server anyway, we'll always log the full startup error on every failed request instead of only once at startup.

**Comment by abernix** [comment]
Using the reproduction this issue was originally opened with (thanks _so_ much for providing that, @NickTomlin!), I can confirm that thanks to changes that have been introduced to Apollo Gateway/Apollo Server's startup, the Gateway fails immediately (at startup) with a clear error message:

```
Error: A valid schema couldn't be composed. The following composition errors were found:
        [accounts] User -> appears to be an entity but no @key directives are specified on the originating type.
    at ApolloGateway.createSchemaFromServiceList (/home/jesse/Dev/apollo/repros/apollo-gateway-key-error/node_modules/@apollo/gateway/dist/index.js:424:19)
    at ApolloGateway.updateByComposition (/home/jesse/Dev/apollo/repros/apollo-gateway-key-error/node_modules/@apollo/gateway/dist/index.js:295:48)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async ApolloGateway.updateSchema (/home/jesse/Dev/apollo/repros/apollo-gateway-key-error/node_modules/@apollo/gateway/dist/index.js:271:13)
    at async ApolloGateway.loadDynamic (/home/jesse/Dev/apollo/repros/apollo-gateway-key-error/node_modules/@apollo/gateway/dist/index.js:250:13)
    at async ApolloGateway.load (/home/jesse/Dev/apollo/repros/apollo-gateway-key-error/node_modules/@apollo/gateway/dist/index.js:221:15)
(Use `node --trace-warnings ...` to show where the warning was created)
```

Closing this with that since this is opened against a specific error in that reproduction, but anyone with a reproduction of a different error should open a new issue with a new reproduction.

Thanks!

**Comment by NickTomlin** [comment]
Thanks for following up on this!
